### PR TITLE
Regex String Functions

### PR DESCRIPTION
**Pull request by gaurav8297**
Functions added:

1. regexp_matches(string, regex)
Returns true if a part of string matches the regex.

2. regexp_replace(string, regex, replacement)
Replaces the first occurrence of regex with the replacement,

3. regexp_extract(string, regex[, group = 0])
Split the string along the regex and extract first occurrence of group.

4. regexp_extract_all(string, regex[, group = 0])
Split the string along the regex and extract all occurrences of group.

I have read and agree to the terms under CLA.md

I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: src/include/common/re2_regex.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    Regex(const std::string &pattern, RegexOptions options = RegexOptions::NONE);
+    const kuzu::regex::RE2 &GetRegex() const {
+        return *regex;
+    }
+
+private:
+    std::shared_ptr<kuzu::regex::RE2> regex;
+};
+
+struct GroupMatch {
```
Comment (acquamarin) [inline src/include/common/re2_regex.h:22]: Better to make the `GroupMatch` and `Match` as class rather than struct

File: src/include/common/re2_regex.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    const std::string &str() const {
+        return text;
+    }
+    operator std::string() const {
+        return text;
+    }
+};
+
+struct Match {
+    std::vector<GroupMatch> groups;
```
Comment (acquamarin) [inline src/include/common/re2_regex.h:35]: Can you add a brief comment about the `Match` and `GroupMatch` class?

File: src/include/common/re2_regex.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <stdexcept>
+#include "re2.h"
+
+namespace kuzu::common {
+enum struct RegexOptions : uint8_t { NONE, CASE_INSENSITIVE };
```
Comment (acquamarin) [inline src/include/common/re2_regex.h:9]: We prefer to use enum class rather than enum struct

File: src/include/common/re2_regex.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    std::vector<GroupMatch> groups;
+
+    GroupMatch &GetGroup(uint64_t index) {
+        if (index >= groups.size()) {
+            throw std::runtime_error("RE2: Match index is out of range");
+        }
+        return groups[index];
+    }
+
+    std::string str(uint64_t index) {
```
Comment (acquamarin) [inline src/include/common/re2_regex.h:44]: Our convention on class method naming is that:
If a method is used to retrive a field from a class, then we name it as: `getXxx`.
E.g. better to rename `position` => `getPosition`.

File: src/include/common/re2_regex.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <stdexcept>
+#include "re2.h"
+
+namespace kuzu::common {
```
Comment (acquamarin) [inline src/include/common/re2_regex.h:8]: ditto, fix the namespace

File: src/function/vector_string_operations.cpp [hunk 6]
```diff
@@ -141,5 +145,59 @@ std::vector<std::unique_ptr<VectorOperationDefinition>> SubStrVectorOperation::g
...
+        TernaryStringExecFunction<ku_string_t, ku_string_t, int64_t, list_entry_t, operation::RegexpExtractAll>,
+        nullptr,
+        bindFunc,
+        false /* isVarLength */));
+    return definitions;
+}
+
+std::unique_ptr<FunctionBindData> RegexpExtractAllOperation::bindFunc(
+    const binder::expression_vector& arguments, FunctionDefinition* definition) {
+    return std::make_unique<FunctionBindData>(DataType(std::make_unique<DataType>(STRING)));
```
Comment (acquamarin) [inline src/function/vector_string_operations.cpp:199]: Should the result type be a ListOfString rather than String?

File: test/test_files/tinysnb/function/string.test [hunk 7]
```diff
@@ -161,17 +161,17 @@ Aida
 Sóló cón tu párejâ
 
 -NAME RegExprEscapeSeq1
--QUERY Return RE_MATCH("peter_n@example.com", ".*\\.com");
```
Comment (acquamarin) [inline test/test_files/tinysnb/function/string.test:164]: Can you group all tests for regex related functions  together?

**Review by acquamarin — CHANGES_REQUESTED** [review]
Good start!
1. Our CI uses the clang-format tool to check the formatting of our code.  You can get a copy of the format checking script here: https://github.com/Sarcasm/run-clang-format/blob/master/run-clang-format.py. Be sure to check the formatting of your code using this tool before pushing.
2. Please make sure that your code can pass all the CI before submitting.
3. We prefer to squash all commits into one for each PR. 
If you have any questions related to my comments, feel free to message me.

File: src/include/common/re2_regex.h [hunk 8]
```diff
@@ -0,0 +1,69 @@
+#pragma once
```
Comment (acquamarin) [inline src/include/common/re2_regex.h:1]: As we discussed, let's remove our wrapper for re2 and use re2 functions directly in our codebase.

File: src/include/function/string/operations/regexp_extract_all_operation.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "common/re2_regex.h"
+#include "common/types/ku_list.h"
+#include "common/types/ku_string.h"
+#include "common/vector/value_vector.h"
```
Comment (acquamarin) [inline src/include/function/string/operations/regexp_extract_all_operation.h:6]: I think value_vector_utils.h includes value_vector and value_vector also includes ku_string.h, so you can probably remove the include for value_vector, ku_string.h, ku_list.h.

File: src/common/re2_regex.cpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                newPosition++;
+            }
+        }
+        position = newPosition;
+        matches.emplace_back(std::move(match));
+    }
+    return matches;
+}
+
+std::string ParseCypherRegex(const std::string& input) {
```
Comment (acquamarin) [inline src/common/re2_regex.cpp:82]: We use camel case for function/variable names. Rename this function to `parseCypherRegex`

File: test/test_files/tinysnb/function/string.test [hunk 11]
```diff
@@ -529,3 +498,79 @@ True
 -QUERY MATCH (p:person) WHERE suffix(p.fName, "l") RETURN p.fName
 ---- 1
 Carol
+
+## Test REGEXP_FULL_MATCHE function
```
Comment (acquamarin) [inline test/test_files/tinysnb/function/string.test:502]: You can remove this comment, as all tests names are self-explanatory

File: src/function/vector_string_operations.cpp [hunk 12]
```diff
@@ -141,5 +135,71 @@ std::vector<std::unique_ptr<VectorOperationDefinition>> SubStrVectorOperation::g
...
+    std::vector<std::unique_ptr<VectorOperationDefinition>> definitions;
+    definitions.emplace_back(make_unique<VectorOperationDefinition>(REGEXP_MATCHES_FUNC_NAME,
+        std::vector<DataTypeID>{STRING, STRING}, BOOL,
+        BinaryExecFunction<ku_string_t, ku_string_t, uint8_t, operation::RegexpMatches>,
+        BinarySelectFunction<ku_string_t, ku_string_t, operation::RegexpMatches>,
+        false /* isVarLength */));
+    return definitions;
+}
+
+std::vector<std::unique_ptr<VectorOperationDefinition>> RegexpReplaceOperation::getDefinitions() {
```
Comment (acquamarin) [inline src/function/vector_string_operations.cpp:159]: Add a TODO here saying that we need to support regexReplace with modifier in the future.

File: src/include/function/string/operations/regexp_extract_all_operation.h [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+namespace kuzu {
+namespace function {
+namespace operation {
+
+struct RegexpExtractAll {
+    static inline void operation(common::ku_string_t& value, common::ku_string_t& pattern,
+        std::int64_t& group, common::list_entry_t& result, common::ValueVector& resultVector) {
+        std::vector<std::string> matches =
+            regexExtractAll(value.getAsString(), pattern.getAsString(), group);
+
```
Comment (acquamarin) [inline src/include/function/string/operations/regexp_extract_all_operation.h:16]: Remove this empty line.

File: src/include/function/string/operations/regexp_extract_all_operation.h [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+        result = common::ListVector::addList(&resultVector, matches.size());
+        auto resultValues = common::ListVector::getListValues(&resultVector, result);
+        auto resultDataVector = common::ListVector::getDataVector(&resultVector);
+        auto numBytesPerValue = resultDataVector->getNumBytesPerValue();
+
```
Comment (acquamarin) [inline src/include/function/string/operations/regexp_extract_all_operation.h:21]: Ditto, let's remove these empty spaces

File: src/include/function/string/operations/regexp_extract_all_operation.h [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        int64_t defaultGroup = 0;
+        operation(value, pattern, defaultGroup, result, resultVector);
+    }
+
+    static std::vector<std::string> regexExtractAll(
+        const std::string& value, const std::string& pattern, std::int64_t& group) {
+        RE2 regex(common::parseCypherPatten(pattern));
+        auto submatchCount = regex.NumberOfCapturingGroups() + 1;
+        if (group > submatchCount) {
+            throw std::runtime_error("RE2: Match index is out of range");
```
Comment (acquamarin) [inline src/include/function/string/operations/regexp_extract_all_operation.h:42]: We have our own exception class, use common::RunTimeException instead of std::runtime_error.

File: src/include/function/string/operations/regexp_extract_all_operation.h [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+        while (regex.Match(input, startPos, input.length(), RE2::UNANCHORED,
+            targetSubMatches.data(), submatchCount)) {
+            uint64_t consumed =
+                static_cast<size_t>(targetSubMatches[0].end() - (input.begin() + startPos));
+            if (!consumed) {
+                // Empty match found, increment the position manually
+                consumed++;
+                while (startPos + consumed < input.length() &&
+                       !IsCharacter(input[startPos + consumed])) {
+                    consumed++;
```
Comment (acquamarin) [inline src/include/function/string/operations/regexp_extract_all_operation.h:60]: Add a test for empty match

File: src/include/common/utils.h [hunk 17]
```diff
@@ -67,5 +68,12 @@ static bool isLittleEndian() {
     return *(uint8_t*)&testNumber == 1;
 }
 
+static inline std::string parseCypherPatten(const std::string& pattern) {
```
Comment (acquamarin) [inline src/include/common/utils.h:71]: I think it is better to have a base regexpOperation class where we put all common methods.
So you can put `parseCypherPatten` and `copyToKuzuString` inside this class.

File: src/include/function/string/operations/regexp_extract_all_operation.h [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+    static inline bool IsCharacter(char c) {
+        // Check if this character is not the middle of utf-8 character i.e. it shouldn't begin with
+        // 10 XX XX XX
+        return (c & 0xc0) != 0x80;
+    }
+
+    static void copyToKuzuString(
+        const std::string& value, common::ku_string_t& kuString, common::ValueVector& valueVector) {
+        if (!common::ku_string_t::isShortString(value.length())) {
+            kuString.overflowPtr = reinterpret_cast<uint64_t>(
```
Comment (acquamarin) [inline src/include/function/string/operations/regexp_extract_all_operation.h:79]: Add a testcase for long string regexp_extract_all

File: src/include/function/string/operations/regexp_extract_operation.h [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+        if (!regex.Match(regex::StringPiece(input), 0, input.length(), RE2::UNANCHORED,
+                targetSubMatches.data(), submatchCount)) {
+            return;
+        }
+
+        copyToKuzuString(targetSubMatches[group].ToString(), result, resultValueVector);
+    }
+
+    static void copyToKuzuString(
```
Comment (acquamarin) [inline src/include/function/string/operations/regexp_extract_operation.h:43]: Move it to the base class.

**Review by acquamarin — APPROVED** [review]
Let's add a base RegexOperation class and let other regexOperations inherit it.
We can put common methods: `parseCypherPattern` and `copyToKuzuString` under this class.

**Review by acquamarin — APPROVED** [review]
Let's add a base RegexOperation class and let other regexOperations inherit it.
We can put common methods: `parseCypherPattern` and `copyToKuzuString` under this class.
